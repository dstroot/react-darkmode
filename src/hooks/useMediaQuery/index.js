import { useState, useEffect } from "react";

const useMediaQuery = query => {
  const [state, setState] = useState(false);

  useEffect(() => {
    let mounted = true;

    // Returns a MediaQueryList object
    const mql = window.matchMedia(query);
    const onChange = () => {
      if (!mounted) return;
      setState(!!mql.matches);
    };

    mql.addEventListener(onChange);
    setState(mql.matches);

    // clean up
    return () => {
      mounted = false;
      mql.removeEventListener(onChange);
    };
  }, [query]);

  return state;
};

export default useMediaQuery;
