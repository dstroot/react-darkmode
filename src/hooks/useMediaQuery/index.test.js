import React from "react";
import { render } from "@testing-library/react";

// import component to test
import useMediaQuery from "../useMediaQuery";
import mediaQuery from "css-mediaquery";

// simulate window resize
const fireResize = width => {
  window.innerWidth = width;
  window.dispatchEvent(new Event("resize"));
};

// mock matchMedia
const createMatchMedia = width => {
  return query => ({
    matches: mediaQuery.match(query, { width }),
    addListener: () => {},
    removeListener: () => {},
    addEventListener: () => {},
    removeEventListener: () => {}
  });
};

// Test component that uses the Hook
const EffecfulComponent = () => {
  const result = useMediaQuery(`(max-width: 400px)`);
  return <span>{result}</span>;
};

describe("useMediaQuery", () => {
  beforeAll(() => {
    window.matchMedia = createMatchMedia(window.innerWidth);
  });

  it("renders", () => {
    const { asFragment } = render(<EffecfulComponent />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("it should change when window width changes", () => {
    fireResize(320);

    const { asFragment } = render(<EffecfulComponent />);
    expect(asFragment()).toMatchSnapshot();
  });
});

// https://medium.com/@dawchihliou/testing-react-hooks-6d3ae95cd838
// https://material-ui.com/components/use-media-query/
