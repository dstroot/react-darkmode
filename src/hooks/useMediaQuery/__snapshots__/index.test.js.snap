// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useMediaQuery it should change when window width changes 1`] = `
<DocumentFragment>
  <span />
</DocumentFragment>
`;

exports[`useMediaQuery renders 1`] = `
<DocumentFragment>
  <span />
</DocumentFragment>
`;
